src/client.ts

```diff
@@ -0,0 +1,217 @@
+    this.handleControl = this.handleControl.bind(this);
+  }
+
+  private async startStream(): Promise<StartStreamResponse> {
+    const response = await fetch(this.config.streamUrl, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        org_id: this.config.orgId,
+        ...this.config.startStreamData,
+      }),
+    });
+
+    if (!response.ok) {
+      throw new Error(`Failed to start stream ${response.statusText}`);
+    }
+
+    return response.json();
+  }
+
+  private buildUrl(token: string, channelId: string): string {
+    const queryParams = new URLSearchParams({
+      token,
+      channel_id: channelId,
+    });
+    if (this.lastEventId) queryParams.set('last_event_id', this.lastEventId);
+    return `${this.config.baseConduitUrl.replace(/\/$/, '')}/events/${this.config.orgId}?${queryParams.toString()}`;
+  }
+
+  private attach(url: string) {
+    if (this.tokenExpiresAt && Date.now() >= this.tokenExpiresAt) {
+      return;
+    }
+
+    this.eventSource = new EventSource(url);
+
+    this.eventSource.onopen = () => {
+      this.config.onOpen?.();
+    };
+
+    if (this.tokenExpiresAt && Date.now() >= this.tokenExpiresAt) {
+      const e: Error & { code?: string } = new Error('Token expired at reconnect');
+      e.code = 'TOKEN_EXPIRED';
+      this.config.onError?.(e);
+      this.disconnect();
+      return;
+    }
+
+    this.eventSource.onerror = () => {
+      if (this.eventSource == null) return;
+      this.config.onError?.(new Error('SSE connection error'));
+    };
+
+    this.eventSource.addEventListener('stream', this.handleStream);
+    this.eventSource.addEventListener('control', this.handleControl);
+  }
+
+  async connect(): Promise<void> {
+    if (this.connecting || this.eventSource) return;
+    this.connecting = true;
+    try {
+      const { token, channel_id, expires_in } = await this.startStream();
+
+      if (channel_id !== this.currentChannelId) {
+        this.currentChannelId = channel_id;
+        this.lastSeq = undefined;
+        this.seenIds.clear();
+      }
+
+      if (Number.isFinite(expires_in) && expires_in > 0) {
+        this.tokenExpiresAt = Date.now() + expires_in * 1000;
+      } else {
+        this.tokenExpiresAt = undefined;
+      }
+
+      const url = this.buildUrl(token, channel_id);
+      this.attach(url);
+    } finally {
+      this.connecting = false;
+    }
+  }
+
+  disconnect(): void {
+    if (this.eventSource) {
+      this.eventSource.onopen = null;
+      this.eventSource.onerror = null;
+      this.eventSource.removeEventListener('stream', this.handleStream);
+      this.eventSource.removeEventListener('control', this.handleControl);
+      this.eventSource.close();
+      this.eventSource = null;
+      this.config.onClose?.();
+    }
+  }
+
+  isConnected(): boolean {
+    return this.eventSource?.readyState === EventSource.OPEN;
+  }
```

Comment:
If you want, you could probably have this discriminate the state of the client, so you're not able to call methods on a client that hasn't connected yet. Here's an example in one of my libraries https://github.com/evanpurkhiser/prolink-connect/blob/main/src/network.ts#L295-L302

In that example some methods / property accessors aren't available until you discriminate on `isConnected`